vm: replace custom result-like types with `Result`

zerbina opened the pull request:
`VmGenResult` and `ExecutionResult` are both kept as aliases.

In addition:
- the handling of the result values is simplified/cleaned up a bit
- errors are always wrapped in `nkError` nodes and no longer reported
 during result unpacking

### Future work:
- `evalStaticStmt`  and `setupCompileTimeVar` need to further propagate
 the errors instead of `localReport`-ing them

zerbina committed:
vm: replace custom result-like types with `Result`

`VmGenResult` and `ExecutionResult` are both kept as aliases.

In addition:
- the handling of the result values is simplified/cleaned up a bit
- errors are always wrapped in `nkError` nodes and no longer reported
 during result unpacking

saem commented on compiler/vm/vm.nim:
```diff
@@ -3319,51 +3304,49 @@ proc execute(c: var TCtx, start: int, frame: sink TStackFrame; cb: proc(c: TCtx,
+  ## Unpacks the vmgen result. If the result represents an error, exits the
+  ## calling function by either returning a new `nkError` wrapping `node`
+  ## or, if the calling function has no return value, via
+  ## `handleReport`+`return`.
```
Can we not use an nkError in both cases except one is an empty?

zerbina replied:
`evalStmt` now also return a `PNode`, making the special `handleReport`+`return` logic here unnecessary.

saem commented on compiler/vm/vm.nim:
```diff
@@ -3319,51 +3304,49 @@ proc execute(c: var TCtx, start: int, frame: sink TStackFrame; cb: proc(c: TCtx,
       result.info = node.info
   else:
-    let errKind = res.report.kind
+    let err = res.takeErr
```
Nonblocking style remark: as you go can you favour let and var blocks?

saem reviewed (commented):
This is super exciting!

We're on the cusp of being able to compile and instead of the current sem approach where we get a boolean, we can instead get the AST that worked out is an nkError!

zerbina commented:
Fixed the failing tests (the VM stack-trace wasn't printed anymore) and also addressed the style remarks.

*Edit:* @saem If the tests pass (they should) and you're okay with the adjustments, feel free to bors already

saem commented:
bors r+

saem commented:
heck yeah, bors away! :D

saem commented:
bors r+
